### PR TITLE
DAOS-6833 build: Build 0.8.0

### PR DESCRIPTION
**Pull request by liw**
Bump the version to 0.8.0. And, set "version-release" for daos-raft* packages.

Signed-off-by: Brian J. Murrell <brian.murrell@intel.com>
Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
The DAOS test build in the pipeline failed:

[2021-05-31T09:42:11.358Z] src/rdb/rdb_raft.c:2607:5: error: void value not ignored as it ought to be
[2021-05-31T09:42:11.358Z]   rc = raft_election_start(db->d_raft);
[2021-05-31T09:42:11.358Z]      ^

which tells us that if we land this, we'll break daos master.

File: debian/changelog [hunk 1]
```diff
@@ -1,3 +1,11 @@
+raft (0.8.0-1) unstable; urgency=medium
+
+  [ Li Wei ]
+  * Add Pre-Vote
+  * Add EL8 support
```
[the remark this answers is not in the document]
Comment (liw) [reply debian/changelog]: Removed.

File: raft.spec [hunk 2]
```diff
@@ -61,6 +61,10 @@ cp -a include/* %{buildroot}/%{_includedir}
 
 
 %changelog
+* Mon May 31 2021 Li Wei <wei.g.li@intel.com> -0.8.0-1
+- Add Pre-Vote
+- Add EL8 support
```
[the remark this answers is not in the document]
Comment (liw) [reply raft.spec]: Removed.

**Comment by liw** [comment]
> The DAOS test build in the pipeline failed:
> 
> [2021-05-31T09:42:11.358Z] src/rdb/rdb_raft.c:2607:5: error: void value not ignored as it ought to be
> [2021-05-31T09:42:11.358Z] rc = raft_election_start(db->d_raft);
> [2021-05-31T09:42:11.358Z] ^
> 
> which tells us that if we land this, we'll break daos master.

@brianjmurrell, could you advise what the official way is, please? I get the feeling I'm probably confused again...

**Comment by brianjmurrell** [comment]
> @brianjmurrell, could you advise what the official way is, please? I get the feeling I'm probably confused again...

So the **best** way is to maintain backward compatibility, if at all possible.  The reason is that while we could change [current master](https://github.com/daos-stack/daos/blob/8b7f761896841919fdc3cfaa7cd6438b51a43813/utils/rpms/daos.spec#L87) such as:

```patch
-BuildRequires: daos-raft-devel >= 0.7.3
+BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 0.8
```

before landing this PR so that the current master will not pick up the new, incompatible (0.8.0) raft, that ship has sailed on past releases which did not specify `daos-raft-devel < 0.8` and trying to rebuild any such release will result in it getting this new incompatible raft.

You might argue that when those past releases were made, they should have had some _upper limit_ (`daos-raft-devel < $something`) also, but of course we can never predict the future (except when we build it into a versioning plan as proposed below) as to when we might want to break compatibility and so at the time of those releases, we didn't know we'd want to break the API in 0.8.0.

If you want to be able to break the API in the future, you can build the ability to do that into your versioning.  You could for example follow the example of dynamic shared libs and make major version changes when the API is broken.  That way any daos release could always have an _upper limit_ being `< next_major_version`.

But again, given that this versioning plan is not in place currently, it would be less-than-deal to break compatibility at this point due to the harm to existing supported releases.

So if you did want to use the _major version updating_ plan, my suggestion would be:
1. land a patch to daos master to specify `BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 1.0`
2. land this PR as raft-1.0.0
3. land a patch to daos master to specify `BuildRequires: daos-raft-devel >= 1.0.0, daos-raft-devel < 2.0` when you update daos to use the new raft API
4. continue to maintain the current raft API[1] until any historic daos release is way out of the supportability window

[1] that means not breaking the API that this PR is proposing to break, but rather moving that breakage into a new API call so that historic releases can still use the existing API.

Or you could just ignore all of that (compatibility with existing releases) and just do the minimum (which is required so that you don't break everyone's PRs in flight when this PR lands) which was first land to daos:

```patch
-BuildRequires: daos-raft-devel >= 0.7.3
+BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 0.8
```

and then land this and then include in your daos PR to use this raft:

```patch
-BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 0.8
-BuildRequires: daos-raft-devel >= 0.8.0
```

but that puts you right back in this position the next time you want to break the API.

**Comment by liw** [comment]
Thanks a lot, @brianjmurrell.

> So the **best** way is to maintain backward compatibility, if at all possible.

I understand. We decide not to, because this is logically an internal module and there's no need for API compatibility (which is always quite a burden and harder than usual for this module).

> The reason is that while we could change [current master](https://github.com/daos-stack/daos/blob/8b7f761896841919fdc3cfaa7cd6438b51a43813/utils/rpms/daos.spec#L87) such as:
> 
> ```diff
> -BuildRequires: daos-raft-devel >= 0.7.3
> +BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 0.8
> ```
> 
> before landing this PR so that the current master will not pick up the new, incompatible (0.8.0) raft, that ship has sailed on past releases which did not specify `daos-raft-devel < 0.8` and trying to rebuild any such release will result in it getting this new incompatible raft.

Looks like we should be fine, as the ">=" is only introduced on master recently:

         master: BuildRequires: daos-raft-devel >= 0.7.3 (we can still change this)
    release/1.2: BuildRequires: raft-devel = 0.7.3
    release/1.0: BuildRequires: raft-devel <= 0.5.0
    release/0.9: BuildRequires: raft-devel <= 0.5.0

Would it work if we change master to

    BuildRequires: daos-raft-devel = 0.7.3

?

> You might argue that when those past releases were made, they should have had some _upper limit_ (`daos-raft-devel < $something`) also, but of course we can never predict the future (except when we build it into a versioning plan as proposed below) as to when we might want to break compatibility and so at the time of those releases, we didn't know we'd want to break the API in 0.8.0.

It's easy to predict. :) We will break the API compatibility again next time we add a new feature to raft. Or actually, I prefer a daos commit builds and statically links with a fixed commit of raft, as that'd be reproducible and be the only combination we test and ship. Then we would have much less to worry about versioning.

> If you want to be able to break the API in the future, you can build the ability to do that into your versioning. You could for example follow the example of dynamic shared libs and make major version changes when the API is broken. That way any daos release could always have an _upper limit_ being `< next_major_version`.
> But again, given that this versioning plan is not in place currently, it would be less-than-deal to break compatibility at this point due to the harm to existing supported releases.

Since RPM packaging was introduced, IIRC, we've been bumping the minor (not major) version whenever we break the API.

> So if you did want to use the _major version updating_ plan, my suggestion would be:
> 
> 1. land a patch to daos master to specify `BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 1.0`
> 2. land this PR as raft-1.0.0
> 3. land a patch to daos master to specify `BuildRequires: daos-raft-devel >= 1.0.0, daos-raft-devel < 2.0` when you update daos to use the new raft API
> 4. continue to maintain the current raft API[1] until any historic daos release is way out of the supportability window
> 
> [1] that means not breaking the API that this PR is proposing to break, but rather moving that breakage into a new API call so that historic releases can still use the existing API.
> 
> Or you could just ignore all of that (compatibility with existing releases) and just do the minimum (which is required so that you don't break everyone's PRs in flight when this PR lands) which was first land to daos:
> 
> ```diff
> -BuildRequires: daos-raft-devel >= 0.7.3
> +BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 0.8
> ```
> 
> and then land this and then include in your daos PR to use this raft:
> 
> ```diff
> -BuildRequires: daos-raft-devel >= 0.7.3, daos-raft-devel < 0.8
> -BuildRequires: daos-raft-devel >= 0.8.0
> ```
> 
> but that puts you right back in this position the next time you want to break the API.

Exactly. Let's keep an upper bound; no open ">=". Would that be OK with you?

**Comment by brianjmurrell** [comment]
> Looks like we should be fine, as the ">=" is only introduced on master recently:
> 
> ```
>      master: BuildRequires: daos-raft-devel >= 0.7.3 (we can still change this)
> release/1.2: BuildRequires: raft-devel = 0.7.3
> release/1.0: BuildRequires: raft-devel <= 0.5.0
> release/0.9: BuildRequires: raft-devel <= 0.5.0
> ```

Ahhh.  Interesting/great detective work there.

> Would it work if we change master to

If the current raft master is 0.7.3 then yes, this should work.

> It's easy to predict. :) We will break the API compatibility again next time we add a new feature to raft.

Without planning though, it's difficult to predict in which version it will happen in, was my point.

> Or actually, I prefer a daos commit builds and statically links with a fixed commit of raft, as that'd be reproducible and be the only combination we test and ship. Then we would have much less to worry about versioning.

Which is not horrible given that statically link raft to daos.

> Since RPM packaging was introduced, IIRC, we've been bumping the minor (not major) version whenever we break the API.

I didn't know that.  Again, though, since we statically link raft it's not as important as it is for shared libs, but the standard (for shared libs at least) is to:
- Bump the major number when you break an existing API's compatibilty
- Bump the minor number when you add new API without breaking existing API
- Bump the bugfix number when you fix a bug that does not have any API impact whatsoever

I just wonder if we should adopt that even though we link statically, even just for our own internal understanding/record-keeping of which raft releases are API compatible and which are not.

> Exactly. Let's keep an upper bound; no open ">=". Would that be OK with you?

Perfectly fine with me.

Thanks for doing the legwork to figure out how we would impact previous releases.  It was very useful.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
Looks fine.  Only -1 until daos master is updated, just to prevent premature landing of this.

**Review by kccain — APPROVED** [review]

**Comment by liw** [comment]
I've created https://github.com/daos-stack/daos/pull/5837 a couple of days ago. It still says "will bump minor version upon API changes", but I'm open to revise this in the future. (We used minor versions probably just to follow the upstream practice or probably I was too shy to have a very large major version.)

**Review by brianjmurrell — APPROVED** [review]

**Comment by liw** [comment]
Synchronized packaging to the content of the latest packaging repository.

**Comment by liw** [comment]
@brianjmurrell, the "test build with DAOS" phase still fails after https://github.com/daos-stack/daos/pull/5837 has landed. Is this because this phase isn't an RPM build? Any thought on how to proceed?

**Comment by brianjmurrell** [comment]
> @brianjmurrell, the "test build with DAOS" phase still fails after [daos-stack/daos#5837](https://github.com/daos-stack/daos/pull/5837) has landed. Is this because this phase isn't an RPM build? Any thought on how to proceed?

It's because it tries to install raft-devel = 0.8.0:

```
[2021-06-15T10:32:13.051Z] DEBUG util.py:446:   raft-devel             x86_64 0.8.0-1388.gca1c84e.el7    raft:PR-47:11    26 k
```

which doesn't make sense because the commit of daos it checked out in that failed stage definitely has:

```
BuildRequires: daos-raft-devel = 0.7.3
```

But putting that aside for the moment (because) what is interesting is that for that test stage, you actually did want the raft-devel-0.8.0 from the PR to be installed, but you want to test with the daos commit/PR that is to be used with this raft 0.8.0 PR.  In order to achieve that, you should add to your commit message in this PR:

```
DAOS-test-branch: liw/raft-prevote
```

(If I have the daos branch correct).  What that does is instructs this PR that when it runs the testing stage, with the new changes in this PR, it should run them with the daos code that's in `liw/raft-prevote`, which is the daos code that supports these changes.

Hopefully that makes sense.

**Comment by liw** [comment]
> But putting that aside for the moment (because) what is interesting is that for that test stage, you actually did want the raft-devel-0.8.0 from the PR to be installed, but you want to test with the daos commit/PR that is to be used with this raft 0.8.0 PR.

Hmm, not quite; I'm actually trying to address your concern that we "don't break everyone's PRs in flight when this PR lands". So I expect the DAOS test build phase to see 0.8.0 but to choose 0.7.3, if it is an RPM build. If it is not an RPM build, I expect it to fail. Perhaps these expectations do not match what the DAOS test build phase does?

The daos PR for liw/raft-prevote will be tested with this raft PR's artifacts to verify that 0.8.0 will work with the updated daos code.

**Comment by brianjmurrell** [comment]
> > But putting that aside for the moment (because) what is interesting is that for that test stage, you actually did want the raft-devel-0.8.0 from the PR to be installed, but you want to test with the daos commit/PR that is to be used with this raft 0.8.0 PR.
> 
> Hmm, not quite; I'm actually trying to address your concern that we "don't break everyone's PRs in flight when this PR lands". So I expect the DAOS test build phase to see 0.8.0 but to choose 0.7.3, if it is an RPM build. If it is not an RPM build, I expect it to fail. Perhaps these expectations do not match what the DAOS test build phase does?
> 
> The daos PR for liw/raft-prevote will be tested with this raft PR's artifacts to verify that 0.8.0 will work with the updated daos code.

So this raises a good point.  A component(/dependency) PR test probably ought to test for both: current-daos-master regression and new-daos-functionality.  But currently it does not and that test stage is intended to test the new functionality (potentially with a new daos PR).

I will look further into why the test stage on this PR is using the new raft PR.

**Comment by liw** [comment]
> So this raises a good point. A component(/dependency) PR test probably ought to test for both: current-daos-master regression and new-daos-functionality. But currently it does not and that test stage is intended to test the new functionality (potentially with a new daos PR).
> 
> I will look further into why the test stage on this PR is using the new raft PR.

Thanks a lot. As to "test the new functionality" from raft.git, I imagine that could involve nontrivial work in Jenkins and CI for raft.git. Since we do cover this testing in corresponding daos.git PRs today, I'd consider that to be some future work, if ever necessary.

**Comment by brianjmurrell** [comment]
Found the problem.  Fixed in daos-stack/raft#48 which is based on this PR so you can just merge that if you want and it will merge into this PR.

**Pull request merged**